File: src/main/java/com/github/rooneyandshadows/lightbulb/annotation_processors/fragment/FragmentScreenGroup.java
package com.github.rooneyandshadows.lightbulb.annotation_processors.fragment;

import com.squareup.javapoet.*;

import javax.lang.model.element.Modifier;
import java.util.ArrayList;

public class FragmentScreenGroup {
    private final String screenGroupName;
    private final ArrayList<FragmentInfo> screens = new ArrayList<>();

    public FragmentScreenGroup(String screenGroupName) {
        if (screenGroupName == null || screenGroupName.equals(""))
            screenGroupName = "Common";
        this.screenGroupName = screenGroupName;
    }

    public String getScreenGroupName() {
        return screenGroupName;
    }

    public ArrayList<FragmentInfo> getScreens() {
        return screens;
    }

    public void addScreen(FragmentInfo fragmentInfo) {
        screens.add(fragmentInfo);
    }

    public TypeSpec build() {
        ClassName baseRouterClass = ClassName.get("com.github.rooneyandshadows.lightbulb.application.activity.routing", "BaseActivityRouter");
        ClassName fragmentScreenClass = baseRouterClass.nestedClass("FragmentScreen");
        TypeSpec.Builder groupClass = TypeSpec
                .classBuilder(screenGroupName)
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL, Modifier.STATIC);
        screens.forEach(fragmentInfo -> {
            boolean hasOptionalParameters = fragmentInfo.hasOptionalParameters();
            MethodSpec.Builder optionalScreenConstructor = MethodSpec
                    .constructorBuilder()
                    .addModifiers(Modifier.PUBLIC);
            MethodSpec.Builder notOptionalScreenConstructor = !hasOptionalParameters ? null : MethodSpec
                    .constructorBuilder()
                    .addModifiers(Modifier.PUBLIC);
            TypeSpec.Builder screenClass = TypeSpec
                    .classBuilder(fragmentInfo.getScreenName())
                    .superclass(fragmentScreenClass)
                    .addModifiers(Modifier.PUBLIC, Modifier.FINAL, Modifier.STATIC);
            MethodSpec.Builder getFragmentMethod = MethodSpec
                    .methodBuilder("getFragment")
                    .addModifiers(Modifier.PUBLIC)
                    .addAnnotation(Override.class)
                    .returns(fragmentInfo.getClassName());
            fragmentInfo.getFragmentParameters().forEach(paramInfo -> {
                String parameterName = paramInfo.name;
                TypeName parameterType = paramInfo.getType();
                FieldSpec.Builder field = FieldSpec.builder(parameterType, parameterName, Modifier.PRIVATE);
                if (paramInfo.isOptional())
                    field.initializer("null");
                screenClass.addField(field.build());
            });
            String allParams = fragmentInfo.generateCommaSeparatedParams(true, paramInfo -> {
                String parameterName = paramInfo.getName();
                optionalScreenConstructor.addParameter(paramInfo.getParameterSpec());
                optionalScreenConstructor.addStatement("this.$L = $L", parameterName, parameterName);
            });
            screenClass.addMethod(optionalScreenConstructor.build());
            if (notOptionalScreenConstructor != null) {
                fragmentInfo.getFragmentParameters(false).forEach(paramInfo -> {
                    String parameterName = paramInfo.getName();
                    if (!paramInfo.isOptional()) {
                        notOptionalScreenConstructor.addParameter(paramInfo.getParameterSpec());
                        notOptionalScreenConstructor.addStatement("this.$L = $L", parameterName, parameterName);
                    }
                });
                screenClass.addMethod(notOptionalScreenConstructor.build());
            }
            getFragmentMethod.addStatement("return $T.newInstance(" + allParams + ")", fragmentInfo.getBindingClassName());
            screenClass.addMethod(getFragmentMethod.build());
            groupClass.addType(screenClass.build());
        });
        return groupClass.build();
    }
}

File: src/main/java/com/github/rooneyandshadows/lightbulb/annotation_processors/fragment/FragmentScreenGroup.kt
package com.github.rooneyandshadows.lightbulb.annotation_processors.fragment

import com.squareup.kotlinpoet.*
import java.util.function.Consumer

class FragmentScreenGroup(screenGroupName: String?) {
    val screenGroupName: String
    val screens = ArrayList<FragmentInfo>()

    init {
        var screenGroupName = screenGroupName
        if (screenGroupName == null || screenGroupName == "") screenGroupName = "Common"
        this.screenGroupName = screenGroupName
    }

    fun addScreen(fragmentInfo: FragmentInfo) {
        screens.add(fragmentInfo)
    }

    fun build(): TypeSpec {
        val baseRouterClass = ClassName(
            "com.github.rooneyandshadows.lightbulb.application.activity.routing",
            "BaseActivityRouter"
        )
        val fragmentScreenClass = baseRouterClass.nestedClass("FragmentScreen")
        val groupClass = TypeSpec.classBuilder(screenGroupName)
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL)
        screens.forEach(Consumer { fragmentInfo: FragmentInfo ->
            val screenConstructor: FunSpec.Builder = FunSpec
                .constructorBuilder()
                .addModifiers(KModifier.PUBLIC)
            val screenClass = TypeSpec.classBuilder(fragmentInfo.screenName!!)
                .superclass(fragmentScreenClass)
                .addModifiers(KModifier.PUBLIC, KModifier.FINAL)
            val getFragmentMethod: FunSpec.Builder = FunSpec
                .builder("getFragment")
                .addModifiers(KModifier.PUBLIC, KModifier.OVERRIDE)
                .addAnnotation(Override::class.java)
                .returns(fragmentInfo.className!!)
            var paramsString = ""
            for (i in fragmentInfo.fragmentParameters.indices) {
                val isLast = i == fragmentInfo.fragmentParameters.size - 1
                val paramInfo = fragmentInfo.fragmentParameters[i]
                val parameterType: TypeName = paramInfo.type
                val parameterName = paramInfo.name
                screenConstructor.addParameter(parameterName, parameterType)
                screenConstructor.addStatement("this.%L = %L", parameterName, parameterName)
                screenClass.addProperty(parameterName, parameterType, KModifier.PRIVATE)
                paramsString = paramsString + if (isLast) parameterName else "$parameterName, "
            }
            getFragmentMethod.addStatement("return %T.newInstance($paramsString)", fragmentInfo.mappedBindingType!!)
            screenClass.addFunction(screenConstructor.build())
            screenClass.addFunction(getFragmentMethod.build())
            groupClass.addType(screenClass.build())
        })
        return groupClass.build()
    }
}